### PR TITLE
Fix(Paywalls) Support Web Views in looping carousels

### PR DESCRIPTION
**Pull request by JZDesign**
### Checklist
- [x] If applicable, unit tests
- [ ] If applicable, create follow-up issues for `purchases-android` and hybrids

### Motivation

If a carousel contained a webview and the carousel was set to loop, it would look like the webview didn't load for a very long time.

### Description

We actually introduced a bug to fix a different bug… And it caused us to only allow one host to own a webview. When we loop carousels, we duplicate the pages. Those pages using the web view all point to the same web view and only one of them can display it. Unfortunately, this meant that the off-screen pages were attempting to render the web view so it wasn't until the loop completed that the web view would be visible at all.

---

> [!NOTE]
> **Medium Risk**
> Changes shared WKWebView attachment and WebKit media suspension across multiple SwiftUI hosts; behavior is well-tested but touches carousel + web view lifecycle edge cases.
> 
> **Overview**
> Fixes **looping carousels** where a shared paywall `WKWebView` could stay attached to an off-screen duplicate page, so the web view looked blank until the carousel finished a full loop.
> 
> **Carousel state** now propagates **`distanceFromActive`** (including nested carousels via `ancestorDistanceFromActive`) instead of `originalCount` / `isActive`, and each carousel page passes that distance into its subtree.
> 
> **Web view hosting** replaces single-host / pending-host logic with **`reconcile`**: track multiple windowed **candidate hosts**, attach the shared web view to the host with the **smallest `carouselDistance`**, and **suspend media** when distance is greater than 1. Media suspend/resume is **deferred to the next main-actor turn** so per-copy SwiftUI updates during a swipe do not flicker playback.
> 
> Tests cover carousel distance behavior, host selection on swipe, tie-breaking, and media suspension.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 24c1a7c9573b0c10be8b163d0077799e247026f8. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: RevenueCatUI/Templates/V2/Components/WebView/WebViewInstance.swift [hunk 1]
```diff
@@ -115,53 +115,66 @@ final class WebViewInstance: ObservableObject {
...
+        guard !self.candidateHosts.contains(where: { $0.host === host }) else {
+            return
         }
 
-        guard self.attachedHost === host else {
+        self.candidateHosts.append(WeakHost(host))
+    }
+
+    /// Moves the web view to the host that should be showing it, or suspends playback when none is.
+    private func reconcileAttachment() {
```
Comment (ajpallares) [inline RevenueCatUI/Templates/V2/Components/WebView/WebViewInstance.swift:136]: Now that the instance knows the carousel distance, should we also use it to gate media playback? If I'm uderstanding correctly, `reconcileAttachment()` only suspends when *no* host is windowed, but in a looping carousel every item stays in the window, so if we have a web view with an `autoplay` video on e.g. page 3, I think it will keep playing audio without the user being able to see it?

Seems like the `VideoComponentView` already considers this using `isActiveOrNeighbor`, so we could use that idea here?

File: RevenueCatUI/Templates/V2/Components/WebView/WebViewInstance.swift [hunk 2]
```diff
@@ -115,53 +115,66 @@ final class WebViewInstance: ObservableObject {
...
         }
 
-        self.attachWebView(to: host)
+        if let attachedHost = self.attachedHost,
+           attachedHost.carouselDistance == closestDistance,
+           candidates.contains(where: { $0 === attachedHost }) {
+            return attachedHost
+        }
+
+        return candidates.first { $0.carouselDistance == closestDistance }
```
Comment (ajpallares) [inline RevenueCatUI/Templates/V2/Components/WebView/WebViewInstance.swift:171]: Hmm one case I'm not sure is covered. With a 1 or 2-page looping carousel, two copies of the *same* page would be visible in the peeks at the same time? I'm wondering whether one of them will show blank

File: RevenueCatUI/Templates/V2/Components/WebView/WebViewComponentView.swift [hunk 3]
```diff
@@ -94,7 +97,8 @@ struct WebViewComponentView: View {
             HostedWebViewComponentView(
                 size: style.size,
                 url: url,
-                instance: instance
+                instance: instance,
+                carouselDistance: self.carouselState?.distanceFromActive ?? 0
```
Comment (ajpallares) [inline RevenueCatUI/Templates/V2/Components/WebView/WebViewComponentView.swift:101]: Probably an edge case, but what if we have a web view inside a carousel nested in an outer *looping* carousel? Wouldn't this overwrite the distance for the inner pages?

File: Tests/RevenueCatUITests/PaywallsV2/WebViewInstanceTests.swift [hunk 4]
```diff
@@ -272,6 +272,89 @@ final class WebViewInstanceHostAttachmentTests: TestCase {
...
+
+        instance.hostDidEnterWindow(activeCopy)
+        instance.hostDidEnterWindow(neighborCopy)
+        instance.updateHost(neighborCopy)
+
+        XCTAssertTrue(webView.superview === activeCopy)
+    }
+
+    /// Swiping changes every copy's distance without any host leaving the window.
+    func testWebViewFollowsTheActivePageWhenTheCarouselMoves() {
```
Comment (ajpallares) [inline Tests/RevenueCatUITests/PaywallsV2/WebViewInstanceTests.swift:312]: This one sets both distances up front and then reconciles twice, but I'm not sure this reproduces the issue being fixed in this PR. I think this would be more faithful to reproduce the origional bug

```swift
firstCopy.carouselDistance = 2
instance.updateHost(firstCopy)   // stale secondCopy still at 2 here
secondCopy.carouselDistance = 0
instance.updateHost(secondCopy)

File: RevenueCatUI/Templates/V2/Components/WebView/WebViewComponentView.swift [hunk 5]
```diff
@@ -179,14 +186,16 @@ struct WebViewRepresentable: PlatformViewRepresentable {
     let url: URL
     let instance: WebViewInstance
     let idStore: WebViewDataStoreIdentifierStore
+    /// See ``WebViewHostView/carouselDistance``.
+    var carouselDistance: Int = 0
```
Comment (ajpallares) [inline RevenueCatUI/Templates/V2/Components/WebView/WebViewComponentView.swift]: Since the only caller relying on this default is a test, could we remove the default value and pass `carouselDistance: 0` explicitly in the test? Should keep us from forgetting to pass it in the future

File: Tests/RevenueCatUITests/PaywallsV2/WebViewInstanceTests.swift
Comment (ajpallares) [inline Tests/RevenueCatUITests/PaywallsV2/WebViewInstanceTests.swift]: There are some tests that still contain the old "PendingHost" naming from before this PR

File: RevenueCatUI/Templates/V2/Components/WebView/WebViewInstance.swift [hunk 6]
```diff
@@ -115,53 +115,66 @@ final class WebViewInstance: ObservableObject {
...
+    }
 
-        self.pendingHost = nil
-        self.attachWebView(to: host)
+    /// Re-evaluates ownership for a host SwiftUI updated while it was already in a window, e.g. because
+    /// its ``WebViewHostView/carouselDistance`` changed when the carousel moved to another page.
+    func updateHost(_ host: WebViewHostView) {
+        self.registerCandidate(host)
+        self.reconcileAttachment()
     }
```
Comment (ajpallares) [inline RevenueCatUI/Templates/V2/Components/WebView/WebViewInstance.swift]: This method is exactly the same as `hostDidEnterWindow`. We shouild probably remove one of them?

File: RevenueCatUI/Templates/V2/EnvironmentObjects/CarouselState.swift [hunk 7]
```diff
@@ -28,6 +28,11 @@ struct CarouselState: Equatable {
     /// The number of original pages (before copies for looping).
     let originalCount: Int
 
+    /// How many pages this page sits from the active one in the data array. `0` for the active page.
+    var distanceFromActive: Int {
```
Comment (ajpallares) [inline RevenueCatUI/Templates/V2/EnvironmentObjects/CarouselState.swift:47]: Could we add a test for this one?

File: RevenueCatUI/Templates/V2/Components/WebView/WebViewComponentView.swift [hunk 8]
```diff
@@ -179,14 +186,16 @@ struct WebViewRepresentable: PlatformViewRepresentable {
...
 
     var expectedOrigin: WebViewOrigin {
         self.instance.session.expectedOrigin
     }
 
     /// One coordinator per instance, retained by the instance: `navigationDelegate` is weak, so a
-    /// coordinator owned by a `ViewThatFits` candidate would stop delivering callbacks the moment
```
Comment (ajpallares) [inline RevenueCatUI/Templates/V2/Components/WebView/WebViewComponentView.swift:188]: I'm curious. Why remove the `ViewThatFits` part of the comment here? (I ask because `ViewThatFits` is still a source of potential duplicated draws). NABD though

**Review by ajpallares — COMMENTED** [review]
Thank you for fixing this! I left some comment though, some of them nits.

File: RevenueCatUI/Templates/V2/Components/WebView/WebViewInstance.swift (hunk 2 above)
Comment (JZDesign) [reply RevenueCatUI/Templates/V2/Components/WebView/WebViewInstance.swift:171]: | 2 item carousel | 1 item carousel |
|:--:|:--:|
|Simulator Screen Recording - iPhone 17 Pro - 2026-09-10 at 09 54 22|Simulator Screen Recording - iPhone 17 Pro - 2026-09-10 at 09 55 23|

File: RevenueCatUI/Templates/V2/Components/WebView/WebViewComponentView.swift (hunk 8 above)
Comment (JZDesign) [reply RevenueCatUI/Templates/V2/Components/WebView/WebViewComponentView.swift:188]: Just because it's any redraw that could cause the original bug, VTF was just the thing that happened to do it when we noticed

Comment (ajpallares) [reply RevenueCatUI/Templates/V2/Components/WebView/WebViewComponentView.swift:188]: Oh right! Thanks for explaining! 🙇

File: RevenueCatUI/Templates/V2/EnvironmentObjects/CarouselState.swift [hunk 9]
```diff
@@ -28,15 +28,34 @@ struct CarouselState: Equatable {
     /// The number of original pages (before copies for looping).
     let originalCount: Int
```
Comment (ajpallares) [inline RevenueCatUI/Templates/V2/EnvironmentObjects/CarouselState.swift:29]: I think this is never read?

File: RevenueCatUI/Templates/V2/EnvironmentObjects/CarouselState.swift [hunk 10]
```diff
@@ -28,15 +28,34 @@ struct CarouselState: Equatable {
...
 
+    /// The effective distance inherited from all enclosing carousels.
+    private let ancestorDistanceFromActive: Int
+
+    init(
+        activeIndex: Int,
+        pageIndex: Int,
+        originalCount: Int,
+        ancestorDistanceFromActive: Int = 0
```
Comment (ajpallares) [inline RevenueCatUI/Templates/V2/EnvironmentObjects/CarouselState.swift:38]: Could we avoid the default parameter here? I think tests can pass 0 if needed

File: RevenueCatUI/Templates/V2/EnvironmentObjects/CarouselState.swift [hunk 11]
```diff
@@ -28,15 +28,34 @@ struct CarouselState: Equatable {
...
+        self.ancestorDistanceFromActive = ancestorDistanceFromActive
+    }
+
+    /// The greatest distance from the active page across this carousel and its ancestors.
+    var distanceFromActive: Int {
+        return max(abs(activeIndex - pageIndex), self.ancestorDistanceFromActive)
+    }
+
     /// Whether this page is the currently visible page in the data array.
     var isActive: Bool {
```
Comment (ajpallares) [inline RevenueCatUI/Templates/V2/EnvironmentObjects/CarouselState.swift:52]: Seems to me like it could happen `isActive == true` while `isActiveOrNeighbor == false`? 

In fact, `isActive` seems to not be called at all, so maybe we can just remove it?

File: RevenueCatUI/Templates/V2/Components/WebView/WebViewComponentView.swift [hunk 12]
```diff
@@ -240,8 +250,10 @@ struct WebViewRepresentable: PlatformViewRepresentable {
 
     @MainActor
     private func update(_ host: WebViewHostView) {
+        // A carousel page change re-updates every copy; the instance needs the new distance before it re-evaluates.
+        host.carouselDistance = self.carouselDistance
         if host.window != nil {
-            self.instance.updateHost(host)
+            self.instance.hostDidEnterWindow(host)
         }
     }
```
Comment (ajpallares) [inline RevenueCatUI/Templates/V2/Components/WebView/WebViewComponentView.swift:258]: Now that suspension is derived from the winner's distance, I think there could be a potential hiccup during swipes. During one swipe we reconcile once per host against a mix of fresh and stale distances, so the intermediate winner can be 2+ pages away and the problem I see is that we would be toggling WebKit several times during the transition.

For example, let's say we have a 3-page looping carousel and the web view on page 0. It copies at data 0/3/6. If we move `index` from 4 to 5, then

- before: distances 4/1/2 --> host3 wins, not suspended
- host0 updates: 5/1/2 --> host3, not suspended
- host3 updates: 5/**2**/2 --> host3 still wins at min distance 2 → **suspended**
- host6 updates: 5/2/**1** → host6 wins → resumed + reparent

So we'd be setting `setAllMediaPlaybackSuspended(true)` and then `(false)` on every swipe past this page if I'm not mistaken. I fear an autoplaying `<video>` may hiccup on this? 

If this does produces hiccups, I wonder if we could somhow coalesce the `setAllMediaPlaybackSuspended` call to avoid this. Cursor proposed something like this

```swift
private var scheduledMediaSuspension: Bool?
private func setMediaPlaybackSuspended(_ suspended: Bool) {
    guard suspended != self.isMediaPlaybackSuspended else {
        // Already in the requested state: drop anything a mid-pass winner queued.
        self.scheduledMediaSuspension = nil
        return
    }
    let wasScheduled = self.scheduledMediaSuspension != nil
    self.scheduledMediaSuspension = suspended
    guard !wasScheduled else { return }
    Task { @MainActor [weak self] in
        self?.flushScheduledMediaSuspension()
    }
}
private func flushScheduledMediaSuspension() {
    guard let suspended = self.scheduledMediaSuspension else { return }
    self.scheduledMediaSuspension = nil
    guard suspended != self.isMediaPlaybackSuspended, let webView = self.webView else { return }
    self.isMediaPlaybackSuspended = suspended
    webView.setAllMediaPlaybackSuspended(suspended)
}
```

Comment (ajpallares) [reply RevenueCatUI/Templates/V2/Components/WebView/WebViewComponentView.swift:258]: Cursor also pointed out that the real cause is that `carouselDistance` is derived state (`activeIndex - pageIndex`) cached per host and refreshed one host at a time. If each host held its stable `CarouselItem.id` plus a reference to a shared per-carousel object with the current active id, `preferredHost()` could compute distances at reconcile time and the first reconcile of a pass would already be correct for every candidate, so there'd be no timing involved at all. 

That would need a class reference inside `CarouselState` though, which complicates `Equatable` and the ancestor composition, so definitely not for this PR...

File: Tests/RevenueCatUITests/PaywallsV2/WebViewInstanceTests.swift [hunk 13]
```diff
@@ -256,22 +261,106 @@ final class WebViewInstanceHostAttachmentTests: TestCase {
...
+
+    func testFartherCarouselCopyCannotTakeTheWebViewFromTheActivePage() {
+        let instance = Self.makeInstance()
+        let webView = instance.webView { WKWebView(frame: .zero) }
+        let activeCopy = self.makeWindowedHost(carouselDistance: 0)
+        let neighborCopy = self.makeWindowedHost(carouselDistance: 1)
+
+        instance.hostDidEnterWindow(activeCopy)
+        instance.hostDidEnterWindow(neighborCopy)
+        instance.hostDidEnterWindow(neighborCopy)
```
Comment (ajpallares) [inline Tests/RevenueCatUITests/PaywallsV2/WebViewInstanceTests.swift:306]: Are these 2 repeeated calls intentional?

File: Tests/RevenueCatUITests/PaywallsV2/WebViewInstanceTests.swift [hunk 14]
```diff
@@ -256,22 +261,106 @@ final class WebViewInstanceHostAttachmentTests: TestCase {
...
+    /// A 2-page loop peeks both neighbours at equal distance; the holder must not bounce between them.
+    func testEquallyDistantCopiesDoNotStealTheWebViewFromEachOther() {
+        let instance = Self.makeInstance()
+        let webView = instance.webView { WKWebView(frame: .zero) }
+        let leftNeighbor = self.makeWindowedHost(carouselDistance: 1)
+        let rightNeighbor = self.makeWindowedHost(carouselDistance: 1)
+
+        instance.hostDidEnterWindow(leftNeighbor)
+        instance.hostDidEnterWindow(rightNeighbor)
+        instance.hostDidEnterWindow(rightNeighbor)
```
Comment (ajpallares) [inline Tests/RevenueCatUITests/PaywallsV2/WebViewInstanceTests.swift:341]: Same question on repeated calls

File: RevenueCatUI/Templates/V2/Components/WebView/WebViewComponentView.swift [hunk 15]
```diff
@@ -240,8 +250,10 @@ struct WebViewRepresentable: PlatformViewRepresentable {
 
     @MainActor
     private func update(_ host: WebViewHostView) {
+        // A carousel page change re-updates every copy; the instance needs the new distance before it re-evaluates.
+        host.carouselDistance = self.carouselDistance
         if host.window != nil {
-            self.instance.updateHost(host)
+            self.instance.hostDidEnterWindow(host)
```
Comment (ajpallares) [inline RevenueCatUI/Templates/V2/Components/WebView/WebViewComponentView.swift:256]: Since we also call `hostDidEnterWindow` from `updateUIView` (where the host didn't actually enter a window), should we use a more neutral name for this? Something like `hostDidUpdate(_:)` or `reconcile(host:)`

**Review by ajpallares — APPROVED** [review]
More comments! Mostly nits, but there's one concern that I think could happen during swipes, since we might be making some contradictory calls to WebKit

Thank you so much for iterating on this! I'm really being nitpicky though. 

I'm approving this because I think that this PR effectively makes the experience better and fixes the original problem, but it'd be great to address this (maybe in a follow-up) to make sure we don't introduce any other issues. Thanks again!
